src/anomalib/metrics/per_image/pimo_numpy.py

```diff
@@ -0,0 +1,461 @@
+
+    See `pimo_curves()` and the module's docstring for more details.
+
+    Args' notation:
+        N: number of images
+        H: image height
+        W: image width
+        K: number of thresholds
+
+    Args:
+        anomaly_maps: floating point anomaly score maps of shape (N, H, W)
+        masks: binary (bool or int) ground truth masks of shape (N, H, W)
+        num_threshs: number of thresholds to compute (K)
+        binclf_algorithm: algorithm to compute the binary classifier curve (see `binclf_curve_numpy.Algorithm`)
+        shared_fpr_metric: metric to compute the shared FPR axis
+        fpr_bounds: lower and upper bounds of the FPR integration range
+        force: whether to force the computation despite bad conditions
+
+    Returns:
+        tuple[ndarray, ndarray, ndarray, ndarray, ndarray]:
+            [0] thresholds of shape (K,) in ascending order
+            [1] shared FPR values of shape (K,) in descending order (indices correspond to the thresholds)
+            [2] per-image TPR curves of shape (N, K), axis 1 in descending order (indices correspond to the thresholds)
+            [3] image classes of shape (N,) with values 0 (normal) or 1 (anomalous)
+            [4] AUPIMO scores of shape (N,) in [0, 1]
+            [5] number of points used in the AUC integration
+    """
+    _validate.rate_range(fpr_bounds)
+
+    # other validations are done in the `pimo` function
+    threshs, shared_fpr, per_image_tprs, image_classes = pimo_curves(
+        anomaly_maps=anomaly_maps,
+        masks=masks,
+        num_threshs=num_threshs,
+        binclf_algorithm=binclf_algorithm,
+        shared_fpr_metric=shared_fpr_metric,
+    )
+    try:
+        _validate.threshs(threshs)
+        _validate.rate_curve(shared_fpr, nan_allowed=False, decreasing=True)
+        _validate.images_classes(image_classes)
+        _validate.per_image_rate_curves(per_image_tprs[image_classes == 1], nan_allowed=False, decreasing=True)
+
+    except ValueError as ex:
+        msg = f"Cannot compute AUPIMO because the PIMO curves are invalid. Cause: {ex}"
+        raise RuntimeError(msg) from ex
+
+    fpr_lower_bound, fpr_upper_bound = fpr_bounds
+
+    # get the threshold indices where the fpr bounds are achieved
+    fpr_lower_bound_thresh_idx, _, fpr_lower_bound_defacto = thresh_at_shared_fpr_level(
+        threshs,
+        shared_fpr,
+        fpr_lower_bound,
+    )
+    fpr_upper_bound_thresh_idx, _, fpr_upper_bound_defacto = thresh_at_shared_fpr_level(
+        threshs,
+        shared_fpr,
+        fpr_upper_bound,
+    )
+
+    if not np.isclose(fpr_lower_bound_defacto, fpr_lower_bound, rtol=(rtol := 1e-2)):
+        msg = (
+            "The lower bound of the shared FPR integration range is not exactly achieved. "
+            f"Expected {fpr_lower_bound} but got {fpr_lower_bound_defacto}, which is not within {rtol=}."
+        )
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
```

Comment:
@ashwinvaidya17 
> What's the advantage of using both logger.warning and warnings.warn. What happens if we want to pipe the warnings to a file and keep the console empty? A file handler can be passed to the logging module in such a scenario.

tbh i didnt know what was the right policy here 😳
i put both cuz i was sure you'd find and tell me what to do :P hehe

so should i use the logger?

on a side note: warnings.warn allows one to make this raise an exception with warnings.filterwarnings("error");
maybe that could be interesting?

for context: this warning is due to having too few points to integrate the AUC curve, which may result in unprecise AUC values